### PR TITLE
Add support for v_exp (exponential)

### PR DESCRIPTION
**Pull request by WanliZhong**
This PR aims to implement `v_exp(v_float16 x)`, `v_exp(v_float32 x)` and `v_exp(v_float64 x)`.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Linux AVX2
```

### CODE REVIEW

**Comment by WanliZhong** [comment]
If there are some ways, can the constant vectors be created only once but used every time?

**Comment by WanliZhong** [comment]
The accuracy of the AVX512 instruction set will be worse when the result is too large.

**Comment by WanliZhong** [comment]
> The accuracy of the AVX512 instruction set will be worse when the result is too large.

The difference comes from the errors of Remes algorithm. Different implements will return different results.

**e^83.4375 =** 
`math.h`: 1.7236370391800008e+36
`ours`: 1.7236371976363258e+36
`other calculator`: 1.7236371016733e+36, 1.7236371016732233e+36, 1.7236371017e+36

All of them are different but the 6 significant bits are the same. So I propose to relax the tests.

**Comment by WanliZhong** [comment]
Need help to identify the error in Android-Test. `Linux AVX2` fail looks like it is related to other PRs which modified the elementwise.

File: modules/core/include/opencv2/core/hal/intrin.hpp [hunk 1]
```diff
@@ -1226,6 +1226,105 @@ namespace CV__SIMD_NAMESPACE {
 //! @endcond
 
 
+//! @name Exponential
+//! @{
+//! @brief Calculates exponent of every vector element
```
Comment (asmorkalov) [inline modules/core/include/opencv2/core/hal/intrin.hpp:1231]: Please add reference to the algorithm with name.

File: modules/core/include/opencv2/core/hal/intrin.hpp [hunk 2]
```diff
@@ -1226,6 +1226,105 @@ namespace CV__SIMD_NAMESPACE {
...
+    inline v_float32 v_exp(const v_float32 &x) {
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/hal/intrin.hpp:1232]: @WanliZhong, we _must_ pay credits to the original author. Please, add the original license for this implementation, as it was done here: https://github.com/vpisarev/ficus/blob/3c9a8b78f49e17489c5e1fd6dd5dd487348c99c2/lib/NN/ficus_nn_common.h#L216. Together with the license, please, put the link to https://github.com/aff3ct/MIPP/blob/master/src/math/neon_mathfun.h

File: modules/core/test/test_intrin_utils.hpp [hunk 3]
```diff
@@ -1794,6 +1794,76 @@ template<typename R> struct TheTest
         return *this;
     }
 
+    TheTest & test_exp_fp32()
+    {
+        int n = VTraits<R>::vlanes();
```
Comment (vpisarev) [inline modules/core/test/test_intrin_utils.hpp:1799]: The test is very weak. v_exp() is critical function. We must test it properly.

1. We must test all the corner and popular cases

```
exp(nan)=nan,
exp(0)=1,
exp(1)~e,
exp(inf)=inf,
exp(-inf)=0,
exp(88) < inf,
exp(89..100)=inf (overflow), // check every number with step 0.1 or even 0.01
0<=exp(-100..-88)<FLT_MIN (underflow). // check every number with step 0.1 or even 0.01
```

2.

```
for (int i = 0; i < 10000; i++) {
    1. generate 4 numbers x_j, j=0..3 in range [-100,100].
    2. compute vector exponent v_exp()
    3. compute exponent for each number using std::exp()
    4. check relative accuracy
         a. std::exp(x_j) = inf => EXPECT_GT(vector_exp_j, FLT_MAX*0.99)
         b. std::exp(x_j) < inf =>
                    EXPECT_GE(vector_exp_j, 0);
                    EXPECT_LT(std::abs(vector_exp_j-std_exp_j) < 1e-6*(std::abs(std_exp_j)+FLT_MIN*100)
```

where 0.99, 1e-6 and 100 should probably be adjusted.

Comment (WanliZhong) [reply modules/core/test/test_intrin_utils.hpp:1799]: Thanks, I will add tests for these corner cases.

**Comment by vpisarev** [comment]
@WanliZhong, I think, we really need to hurry up. Please, finalize this PR as soon as possible. Then, submit another PR with v_log, v_sin and v_cos implemented.

File: modules/core/include/opencv2/core/hal/intrin.hpp (hunk 1 above)
Comment (opencv-alalek) [inline modules/core/include/opencv2/core/hal/intrin.hpp:1231]: Almost all documentation in this file is ignored.

This file contains too much conditional compilations through `#if` and template-based logic.
Doxygen doesn't work properly with that.

Documentation is stored and well structured in intrin_cpp.hpp

File: modules/core/include/opencv2/core/hal/intrin.hpp [hunk 4]
```diff
@@ -1226,6 +1226,105 @@ namespace CV__SIMD_NAMESPACE {
...
+        const v_float32 _vexp_lo_f32 = vx_setall_f32(-88.3762626647949f);
+        const v_float32 _vexp_hi_f32 = vx_setall_f32(89.f);
+        const v_float32 _vexp_half_fp32 = vx_setall_f32(0.5f);
+        const v_float32 _vexp_one_fp32 = vx_setall_f32(1.f);
+        const v_float32 _vexp_LOG2EF_f32 = vx_setall_f32(1.44269504088896341f);
```
Comment (opencv-alalek) [inline modules/core/include/opencv2/core/hal/intrin.hpp:1237]: 1.
My suggestion is to move this and similar implementations into a separate file.
There is already `simd_utils.impl.hpp`.
It could be reused or created a new one like `simd_math.impl.hpp`

2.
Also all functions should be guarded if some ISA has better support for these functions.
At least intrin_cpp.hpp could use `std::exp()` as reference (to ensure that test code is adequate)

**Review by opencv-alalek — COMMENTED** [review]
Take a look how to run SIMD emulator configuration (intrin_cpp.hpp):

https://pullrequest.opencv.org/buildbot/builders/4_x_etc-simd-emulator-lin64/builds/100378

File: modules/core/test/test_intrin_utils.hpp (hunk 3 above)
Comment (WanliZhong) [reply modules/core/test/test_intrin_utils.hpp:1799]: since `v_max(nan, x)` and `v_min(nan, x)` will not return `nan` except neon, I disabled this test. If there are some way to handle this cases?

File: modules/core/include/opencv2/core/hal/intrin.hpp (hunk 1 above)
Comment (WanliZhong) [reply modules/core/include/opencv2/core/hal/intrin.hpp]: OK, I added a line in `intrin_cpp.hpp` for document
```
| Operations\\Types | float 32 | float 64 |
|-------------------|:-:|:-:|
|exp                | x | x |
```

File: modules/core/include/opencv2/core/hal/intrin.hpp (hunk 4 above)
Comment (WanliZhong) [reply modules/core/include/opencv2/core/hal/intrin.hpp:1265]: 1. The constants will only used in the `v_exp` function currently. Maybe some of them will be used in other functions like `v_sin` and `v_cos`? Should I move them to a separate file as global variables? I just want them to be created when they are used for the first time.
2. There is already `OPENCV_HAL_IMPL_MATH_FUNC(v_exp, std::exp, _Tp)` in line 728 of file `intrin_cpp.hpp `

Comment (asmorkalov) [reply modules/core/include/opencv2/core/hal/intrin.hpp:1265]: Let's implement item 2 too. I want to benchmark both options before merge.

Comment (asmorkalov) [reply modules/core/include/opencv2/core/hal/intrin.hpp:1265]: Also I propose to add reference to some paper, description, or implementation details, if you have it.

Comment (WanliZhong) [reply modules/core/include/opencv2/core/hal/intrin.hpp:1265]: > Also I propose to add reference to some paper, description, or implementation details, if you have it.

I cannot find the original paper; I only found the code and some comments. I already added these comments and references.

**Comment by fengyuentau** [comment]
Possible to merge it soon? This is needed by v_erf, which is needed by gelu acceleration.

File: modules/core/include/opencv2/core/hal/intrin.hpp (hunk 4 above)
Comment (asmorkalov) [reply modules/core/include/opencv2/core/hal/intrin.hpp:1265]: The fallback with just loop and standatrd exp is still needed. We have platforms, where fma/sse/avx/neon are not implemented. Looped exp should be better there.

File: modules/core/include/opencv2/core/hal/intrin_cpp.hpp [hunk 5]
```diff
@@ -721,10 +722,28 @@ template<typename _Tp, int n> inline v_reg<_Tp2, n> func(const v_reg<_Tp, n>& a)
 Only for floating point types.*/
 OPENCV_HAL_IMPL_MATH_FUNC(v_sqrt, std::sqrt, _Tp)
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/hal/intrin_cpp.hpp:724]: insert 

```
#define OPENCV_HAL_MATH_HAVE_EXP 1
```

**Comment by WanliZhong** [comment]
@asmorkalov Hi Alexander, I think we can review this PR again. I have finalized the code as the comments that you, Vadim and Alekin left before.

**Review by vpisarev — APPROVED** [review]

**Review by asmorkalov — APPROVED** [review]

**Comment by opencv-alalek** [comment]
There is regression detected by weekly build: http://pullrequest.opencv.org/buildbot/builders/4_x_etc-simd-emulator-lin64/builds/100410

**Comment by asmorkalov** [comment]
I was able to reproduce the regression locally on Linux host with `cmake -DOPENCV_EXTRA_FLAGS="-DCV_FORCE_SIMD128_CPP=1" ../opencv-master
`

**Comment by WanliZhong** [comment]
I have reproduced this error too, I am trying to fix it

File: modules/core/test/test_intrin_utils.hpp [hunk 6]
```diff
@@ -2035,13 +2133,13 @@ void test_hal_intrin_float64()
         .test_mask()
         .test_unpack()
         .test_float_math()
-        .test_round_pair_f64()
```
[the remark this answers is not in the document]
Comment (WanliZhong) [reply modules/core/test/test_intrin_utils.hpp:2038]: I think it's by accident. I will add it back.

File: modules/core/include/opencv2/core/hal/intrin_math.hpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        _vexp_y = v_mul(_vexp_y, v_reinterpret_as_f16(_vexp_mm));
+
+        // exp(NAN) -> NAN
+        v_float16 mask_not_nan = v_not_nan(x);
+        return v_select(mask_not_nan, _vexp_y, v_reinterpret_as_f16(vx_setall_s16(0x7e00)));
+    }
+#endif
+
+    inline v_float32 v_exp(const v_float32 &x) {
+        const v_float32 _vexp_lo_f32 = vx_setall_f32(-88.3762626647949f);
```
[the remark this answers is not in the document]
Comment (WanliZhong) [reply modules/core/include/opencv2/core/hal/intrin_math.hpp:94]: So should I provide `v_`, `v256_`, `v512_` and `vx_` versions implementation in `intrin_math.hpp` like `simd_utils.impl.hpp`? That's really will make many duplicated code.

[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/core/include/opencv2/core/hal/intrin_math.hpp:94]: It must be a better way of doing it without duplication.
